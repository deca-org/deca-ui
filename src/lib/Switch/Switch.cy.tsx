import { Test } from '../Utils';
import _cyp from '../../../cypress';
import React from 'react';
import Switch from './Switch';
import { standardColors } from '../Theme';

describe('components/Switch', () => {
  describe('base', () => {
    describe('colors', () => {
      standardColors.map((color) =>
        it(color, () => {
          cy.mount(<Switch color={color} />);
          cy.get('label')
            .before('background-color')
            .should('eq', Test.color('gray300'));

          cy.get('label').click();

          cy.wait(250);

          cy.get('label')
            .before('background-color')
            .should('eq', Test.color(color));

          cy.get('label')
            .after('background-color')
            .should('eq', Test.color('white'));
        })
      );
    });
    describe('disabled', () => {
      it('unselected', () => {
        cy.mount(<Switch disabled />);
        cy.get('label')
          .before('background-color')
          .should('eq', Test.color('gray200'));

        cy.get('label')
          .after('background-color')
          .should('eq', Test.color('gray400'));
      });
      it('selected', () => {
        cy.mount(<Switch initialToggle disabled />);
        cy.get('label');
        cy.get('label')
          .before('background-color')
          .should('eq', Test.color('gray200'));

        cy.get('label')
          .after('background-color')
          .should('eq', Test.color('gray400'));
      });
    });
  });
  describe('dark mode', () => {
    describe('colors', () => {
      standardColors.map((color) =>
        it(color, () => {
          cy.darkMount(<Switch color={color} />);
          cy.get('label')
            .before('background-color')
            .should('eq', Test.color('gray800'));

          cy.get('label').click();

          cy.wait(250);

          cy.get('label')
            .before('background-color')
            .should('eq', Test.color(color));

          cy.get('label')
            .after('background-color')
            .should('eq', Test.color('black'));
        })
      );
    });
    describe('disabled', () => {
      it('unselected', () => {
        cy.darkMount(<Switch disabled />);
        cy.get('label')
          .before('background-color')
          .should('eq', Test.color('gray700'));

        cy.get('label')
          .after('background-color')
          .should('eq', Test.color('gray600'));

        cy.get('label').after('opacity').should('eq', '0.4');
      });
      it('selected', () => {
        cy.darkMount(<Switch initialToggle disabled />);
        cy.get('label')
          .before('background-color')
          .should('eq', Test.color('gray700'));

        cy.get('label')
          .after('background-color')
          .should('eq', Test.color('gray600'));

        cy.get('label').after('opacity').should('eq', '0.4');
      });
    });
  });
});
