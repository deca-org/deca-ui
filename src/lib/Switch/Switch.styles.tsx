import { styled } from '@lib/Theme/stitches.config';

export const StyledSwitchWrapper = styled('div', {
  boxSizing: 'border-box',
  display: 'flex',
  alignItems: 'center',
  fontFamily: '$normal',
  transition: '$default',
});

export const StyledSwitchInput = styled('input', {
  opacity: 0,
  position: 'absolute',
  cursor: 'pointer',
  '&:focus-visible + label::before': {
    boxShadow: '$a11y',
  },
  compoundVariants: [
    {
      isDark: true,
      isDisabled: true,
      css: {
        '&:checked + label::before': {
          bg: '$gray700',
        },
        '&:checked:active + label::before': {
          bg: '$gray700',
        },
      },
    },
    {
      isDark: false,
      isDisabled: true,
      css: {
        '&:checked + label::before': {
          bg: '$gray200',
        },
        '&:checked:active + label::before': {
          bg: '$gray200',
        },
      },
    },
  ],
  variants: {
    size: {
      sm: {
        '&:checked + label::after': {
          transform: 'translateX(calc(100% + $space$0))',
        },
        '&:active + label::after': {
          pr: '$1',
        },
        '&:checked:active + label::after': {
          transform: 'translateX(calc(100% - ($space$1 + $space$0)))',
        },
      },
      md: {
        '&:checked + label::after': {
          transform: 'translateX(calc(100% - $space$0))',
        },
        '&:active + label::after': {
          pr: '$1',
        },
        '&:checked:active + label::after': {
          transform: 'translateX(calc(100% - ($space$0 + $space$2)))',
        },
      },
      lg: {
        '&:checked + label::after': {
          transform: 'translateX(calc(100% - $space$1))',
        },
        '&:active + label::after': {
          pr: '$1',
        },
        '&:checked:active + label::after': {
          transform: 'translateX(calc(100% - ($space$1 + $space$2)))',
        },
      },
    },
    color: {
      primary: {
        '&:checked + label::before': {
          bg: '$primary',
        },
        '&:checked:active + label::before': {
          bg: '$primary-darken-1',
        },
      },
      secondary: {
        '&:checked + label::before': {
          bg: '$secondary',
        },
        '&:checked:active + label::before': {
          bg: '$secondary-darken-1',
        },
      },
      success: {
        '&:checked + label::before': {
          bg: '$success',
        },
        '&:checked:active + label::before': {
          bg: '$success-darken-1',
        },
      },
      warning: {
        '&:checked + label::before': {
          bg: '$warning',
        },
        '&:checked:active + label::before': {
          bg: '$warning-darken-1',
        },
      },
      error: {
        '&:checked + label::before': {
          bg: '$error',
        },
        '&:checked:active + label::before': {
          bg: '$error-darken-1',
        },
      },
    },
    isDisabled: {
      true: {},
      false: {},
    },
    isDark: {
      true: {},
      false: {},
    },
  },
});

export const StyledSwitchLabel = styled('label', {
  color: '$text',
  position: 'relative',
  boxSizing: 'content-box',
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  '&::before': {
    boxSizing: 'content-box',
    content: '',
    br: '$pill',
    transition: '$default',
  },
  '&::after': {
    boxSizing: 'content-box',
    position: 'absolute',
    content: '',
    br: '$pill',
    transition: '$default',
  },
  compoundVariants: [
    {
      hasLabel: true,
      size: 'sm',
      css: {
        '&::before': {
          mr: '$1',
        },
      },
    },
    {
      hasLabel: true,
      size: 'md',
      css: {
        '&::before': {
          mr: '$2',
        },
      },
    },
    {
      hasLabel: true,
      size: 'lg',
      css: {
        '&::before': {
          mr: '$2',
        },
      },
    },
    {
      isDark: true,
      isDisabled: true,
      css: {
        '&::before': {
          bg: '$gray700',
        },
        '&::after': {
          opacity: '40%',
          bg: '$gray600',
        },
      },
    },
    {
      isDark: false,
      isDisabled: true,
      css: {
        '&::before': {
          bg: '$gray200',
        },
        '&::after': {
          bg: '$gray400',
        },
      },
    },
  ],
  variants: {
    size: {
      sm: {
        fontSize: '$caption',
        '&::before': {
          width: '$7',
          height: '$3',
        },
        '&::after': {
          left: 'calc($0 + $space$1 - $space$1)',
          size: 'calc($4 - $space$2)',
        },
      },
      md: {
        fontSize: '$bodySm',
        '&::before': {
          width: '$8',
          height: '$4',
        },
        '&::after': {
          left: 'calc($0 + $space$1 - $space$0 / 2)',
          size: 'calc($4 - $space$1 - $space$0)',
        },
      },
      lg: {
        fontSize: '$body',
        '&::before': {
          width: '$9',
          height: '$5',
        },
        '&::after': {
          left: 'calc($0 + $space$0)',
          size: 'calc($4 - $space$0)',
        },
      },
    },
    color: {
      primary: {},
      secondary: {},
      success: {},
      warning: {},
      error: {},
    },
    isDisabled: {
      true: {
        cursor: 'not-allowed',
      },
      false: {},
    },
    hasLabel: {
      true: {},
      false: {},
    },
    isDark: {
      true: {
        '&::before': {
          bg: '$gray800',
        },
        '&::after': {
          bg: '$black',
        },
      },
      false: {
        '&::before': {
          bg: '$gray300',
        },
        '&::after': {
          bg: '$white',
        },
      },
    },
  },
});
