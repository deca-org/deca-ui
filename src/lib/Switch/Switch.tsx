import { ThemeContext } from '@lib/Theme';
import { CSS, StandardColors } from '@lib/Theme/stitches.config';
import {
  Modify,
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
  uuid,
  __DEV__,
} from '@lib/Utils';
import clsx from 'clsx';
import React, { useState, useMemo } from 'react';

import {
  StyledSwitchWrapper,
  StyledSwitchInput,
  StyledSwitchLabel,
} from './Switch.styles';
/**
 * Switches are an alternative to the checkbox component. You can switch between enabled or disabled states.
 */
interface Props
  extends Modify<
    React.ComponentPropsWithRef<'input'>,
    {
      /**
       * Size of the component.
       * @default md
       */
      size?: 'sm' | 'md' | 'lg';
    }
  > {
  /**
   * Size of the component.
   * @default md
   */
  size?: 'sm' | 'md' | 'lg';
  /**
   * Text label for the component.
   */
  label?: string;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * Disabled state applied to the component.
   * @default false
   */
  disabled?: boolean;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * Color of radio button when active.
   * @default primary
   */
  color?: StandardColors;
  /**
   * The value assigned to the switch component.
   */
  value?: string;
  /**
   * Callback fired when a switch is enabled or disabled
   */
  onChange?(e: React.ChangeEvent<HTMLInputElement>): void;
  /**
   * Callback fired when a switch is focused
   */
  onFocus?(e: React.ChangeEvent<HTMLInputElement>): void;
  /**
   * Name is used as an identifier in a form.
   */
  name?: string;
  /**
   *  Whether or not the switch is initially toggled.
   */
  initialToggle?: boolean;
  /**
   * Whether or not the switch is toggled.
   */
  toggled?: boolean;
  /**
   * Required checkbox prop.
   * @default false
   */
  required?: boolean;
}

export type SwitchProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type SwitchComponent = (<C extends React.ElementType = 'input'>(
  props: SwitchProps<C>
) => React.ReactElement | null) & { displayName?: string };

const Switch: SwitchComponent = React.forwardRef(
  <T extends React.ElementType = 'input'>(
    {
      size = 'md',
      label,
      className,
      disabled = false,
      as,
      css,
      color = 'primary',
      required = false,
      toggled,
      initialToggle = false,
      onChange,
      name,
      value,
    }: SwitchProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const [selfToggled, setSelfToggled] = useState<boolean>(initialToggle);

    const switchId = uuid('switch');

    const isControlledComponent = useMemo(
      () => toggled !== undefined,
      [toggled]
    );

    const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
      if (disabled) return;
      if (!isControlledComponent) {
        setSelfToggled(e.target.checked);
      }
      onChange && onChange(e);
    };

    const preClass = 'decaSwitch';

    const { dark } = React.useContext(ThemeContext);

    return (
      <StyledSwitchWrapper className={clsx(className, `${preClass}-root`)}>
        <StyledSwitchInput
          id={switchId}
          type="checkbox"
          onChange={changeHandler}
          checked={isControlledComponent ? toggled : selfToggled}
          ref={ref}
          isDisabled={disabled}
          name={name}
          size={size}
          className={`${preClass}-input`}
          disabled={disabled}
          color={color}
          required={required}
          value={value}
          isDark={dark}
        />
        <StyledSwitchLabel
          htmlFor={switchId}
          css={css}
          size={size}
          as={as as any} // as prop will not accept T in this case
          color={color}
          isDisabled={disabled}
          hasLabel={label ? true : false}
          isDark={dark}
        >
          {label && label}
        </StyledSwitchLabel>
      </StyledSwitchWrapper>
    );
  }
);

if (__DEV__) {
  Switch.displayName = 'DecaUI.Switch';
}

export default Switch;
