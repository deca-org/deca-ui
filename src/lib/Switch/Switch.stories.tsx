import { DecaUIProvider } from '@lib/Theme';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import Switch from './Switch';

export default {
  title: 'Switch',
  component: Switch,
} as ComponentMeta<typeof Switch>;

const Template: ComponentStory<typeof Switch> = (args) => <Switch {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Label',
  size: 'md',
  color: 'primary',
  disabled: false,
  css: {},
  className: '',
  initialToggle: true,
  required: false,
};

export const NoLabel = Template.bind({});
NoLabel.args = {
  ...Default.args,
  label: '',
};

export const WithTheme = Template.bind({});

WithTheme.args = { ...Default.args };
WithTheme.decorators = [
  (Story) => (
    <DecaUIProvider
      theme={{
        colors: {
          primary: '$green600',
        },
      }}
    >
      <Story />
    </DecaUIProvider>
  ),
];

export const DarkMode = Template.bind({});

DarkMode.args = { ...Default.args };
DarkMode.parameters = { backgrounds: { default: 'dark' } };

DarkMode.decorators = [
  (Story) => (
    <DecaUIProvider mode="dark">
      <Story />
    </DecaUIProvider>
  ),
];
