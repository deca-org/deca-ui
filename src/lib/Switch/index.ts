import Switch from './Switch';
export * from './Switch';
export { Switch };
export default Switch;
