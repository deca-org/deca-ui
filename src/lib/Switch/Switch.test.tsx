import { render, act } from '@testing-library/react';
import React from 'react';

import Switch from './Switch';

describe('components/Switch', () => {
  it('matches screenshot', () => {
    const { asFragment } = render(<Switch label="Label Text" />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('onChange event fires', () => {
    const mockFn = jest.fn();
    const utils = render(<Switch label="Label Text" onChange={mockFn} />);
    const input = utils.getByLabelText('Label Text');
    act(() => {
      input.click();
    });
    expect(mockFn.mock.calls.length).toBe(1);
  });
  it('should ignore events when disabled', () => {
    const mockFn = jest.fn();
    const utils = render(
      <Switch label="Label Text" onChange={mockFn} disabled />
    );
    const input = utils.getByLabelText('Label Text');
    act(() => {
      input.click();
    });
    expect(mockFn.mock.calls.length).toBe(0);
  });
  it('renders all colors', () => {
    const { asFragment } = render(
      <>
        <Switch color="primary" />
        <Switch color="secondary" />
        <Switch color="success" />
        <Switch color="warning" />
        <Switch color="error" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders all sizes', () => {
    const { asFragment } = render(
      <>
        <Switch size="sm" />
        <Switch size="md" />
        <Switch size="lg" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
