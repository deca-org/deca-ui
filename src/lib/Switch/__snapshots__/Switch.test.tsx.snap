// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Switch matches screenshot 1`] = `
<DocumentFragment>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-ejeiXn-size-md c-eZVbbP-esaJkh-color-primary decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-eFiyFa-size-md c-dhssrN-TIRgJ-isDark-false c-dhssrN-jQVVMl-cv"
      for="switch-1"
    >
      Label Text
    </label>
  </div>
</DocumentFragment>
`;

exports[`components/Switch renders all colors 1`] = `
<DocumentFragment>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-ejeiXn-size-md c-eZVbbP-esaJkh-color-primary decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-eFiyFa-size-md c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-ejeiXn-size-md c-eZVbbP-gfACsx-color-secondary decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-eFiyFa-size-md c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-ejeiXn-size-md c-eZVbbP-jYLLl-color-success decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-eFiyFa-size-md c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-ejeiXn-size-md c-eZVbbP-dKPMtd-color-warning decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-eFiyFa-size-md c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-ejeiXn-size-md c-eZVbbP-hPZWBd-color-error decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-eFiyFa-size-md c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
</DocumentFragment>
`;

exports[`components/Switch renders all sizes 1`] = `
<DocumentFragment>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-jWLXHG-size-sm c-eZVbbP-esaJkh-color-primary decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-kNuBnU-size-sm c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-ejeiXn-size-md c-eZVbbP-esaJkh-color-primary decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-eFiyFa-size-md c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
  <div
    class="c-gidqah decaSwitch-root"
  >
    <input
      class="c-eZVbbP c-eZVbbP-hPARGX-size-lg c-eZVbbP-esaJkh-color-primary decaSwitch-input"
      id="switch-1"
      type="checkbox"
      value=""
    />
    <label
      class="c-dhssrN c-dhssrN-iXIkLl-size-lg c-dhssrN-TIRgJ-isDark-false"
      for="switch-1"
    />
  </div>
</DocumentFragment>
`;
